### PR TITLE
Add an iris-esmf-regrid based regridding scheme

### PR DESCRIPTION
**Pull request by bouweandela**
## Description

Add an iris-esmf-regrid based regridding scheme

Part of #2405.

Link to documentation:

***

## [Before you get started](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#getting-started)

- [x] [☝ Create an issue](https://github.com/ESMValGroup/ESMValCore/issues) to discuss what you are going to do

## [Checklist](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#checklist-for-pull-requests)

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [ ] [🧪][2] The new functionality is [relevant and scientifically sound](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#scientific-relevance)
- [ ] [🛠][1] This pull request has a [descriptive title and labels](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-title-and-label)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#code-quality)
- [x] [🧪][2] and [🛠][1] [Documentation](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#documentation) is available
- [ ] [🛠][1] [Unit tests](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#tests) have been added
- [x] [🛠][1] Changes are [backward compatible](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#backward-compatibility)
- [x] [🛠][1] The [list of authors](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#list-of-authors) is up to date
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/contributing.html#pull-request-checks) were successful

***

To help with the number pull requests:

- 🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValCore/pulls) in this repository

### CODE REVIEW

File: esmvalcore/preprocessor/_regrid.py [hunk 1]
```diff
@@ -533,9 +535,13 @@ def _get_target_grid_cube(
...
 
-def _attempt_irregular_regridding(cube: Cube, scheme: str) -> bool:
+def _attempt_irregular_regridding(
+    src_cube: Cube,
+    tgt_cube: Cube,
+    scheme: str,
+) -> bool:
     """Check if irregular regridding with ESMF should be used."""
-    if not has_irregular_grid(cube):
+    if not (has_irregular_grid(src_cube) or has_irregular_grid(tgt_cube)):
```
Comment (bouweandela) [inline esmvalcore/preprocessor/_regrid.py:544]: @schlunma Does this also work for ugrid data?

File: doc/recipe/preprocessor.rst [hunk 2]
```diff
@@ -997,9 +1001,12 @@ with the remaining entries of the ``scheme`` dictionary passed as keyword
 arguments.
 
 One package that aims to capitalize on the :ref:`support for unstructured grids
-introduced in Iris 3.2 <iris:ugrid>` is :doc:`iris-esmf-regrid:index`.
+introduced in Iris 3.2 <iris:ugrid>` is :doc:`esmf_regrid:index`.
 It aims to provide lazy regridding for structured regular and irregular grids,
```
[the remark this answers is not in the document]
Comment (schlunma) [reply doc/recipe/preprocessor.rst:1005]: Are you referring to "our" definition of "irregular" (i.e., general curvilinear grids)? Yes, I think it does, since most likely some of the models [here](https://esmvaltool.dkrz.de/shared/esmvaltool/regrid_tests/recipe_tos_regrid_20240614_105846/) use that type of grid.

File: esmvalcore/preprocessor/_regrid.py (hunk 1 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_regrid.py:544]: Yes!

```py
>>> # cube is a UGRID-compliant cube
>>> ih.has_regular_grid(cube)
False
>>> ih.has_irregular_grid(cube)
False
>>> ih.has_unstructured_grid(cube)
True
```

Please note that in the UGRID word "irregular" and "unstructured" might mean the same, whereas for us "irregular" means "curvilinear"

File: doc/conf.py [hunk 3]
```diff
@@ -438,20 +438,21 @@
...
 # Configuration for intersphinx
 intersphinx_mapping = {
-    'cf_units': ('https://cf-units.readthedocs.io/en/latest/', None),
+    'cf_units': ('https://cf-units.readthedocs.io/en/stable/', None),
     'cftime': ('https://unidata.github.io/cftime/', None),
     'esmvalcore':
     (f'https://docs.esmvaltool.org/projects/ESMValCore/en/{rtd_version}/',
      None),
     'esmvaltool': (f'https://docs.esmvaltool.org/en/{rtd_version}/', None),
+    "esmpy": ("https://earthsystemmodeling.org/esmpy_doc/release/latest/html/",
```
Comment (schlunma) [inline doc/conf.py:447]:
```suggestion
    'esmpy': ('https://earthsystemmodeling.org/esmpy_doc/release/latest/html/',
```

Just to be consistent with the other entries

File: esmvalcore/preprocessor/_regrid.py [hunk 4]
```diff
@@ -39,10 +42,9 @@
     add_cell_measure,
 )
 from esmvalcore.preprocessor.regrid_schemes import (
-    ESMPyAreaWeighted,
-    ESMPyLinear,
-    ESMPyNearest,
     GenericFuncScheme,
+    ESMPyNearest,
+    IrisESMFRegrid,
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid.py:47]: Run `isort`?

File: esmvalcore/preprocessor/_regrid.py (hunk 1 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid.py:544]: But for UGRID cubes we would like to use iris-esmf-regrid. What would be the best way to proceed? Add another function that checks if the source or target cube has a `iris.cube.Cube.mesh` property that is not `None` and that case use iris-esmf-regrid?

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+    def __init__(
+        self,
+        method: Literal["bilinear", "conservative", "nearest"],
+        mdtol: float | None = None,
+        use_src_mask: bool | np.ndarray = True,
+        use_tgt_mask: bool | np.ndarray = True,
+        src_resolution: int | None = None,
+        tgt_resolution: int | None = None,
+        tgt_location: Literal['face', 'node'] | None = None,
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:102]: Would it maybe make sense to pass all of these as `**kwargs`? It's nice to have them explicitly documented here, but if the regridders support new arguments in the future or change arguments we will need to adapt this code here to support this.

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            mask = da.ma.getmaskarray(data[slices])
+        else:
+            vertical_dim = cube.coord_dims(vertical_coord)[0]
+            slices = tuple(
+                slice(None) if i in horizontal_dims + (vertical_dim, ) else 0
+                for i in range(cube.ndim))
+            mask = da.ma.getmaskarray(data[slices])
+            mask_vertical_dim = sum(i < vertical_dim for i in horizontal_dims)
+            mask = mask.all(axis=mask_vertical_dim)
+        return mask
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:168]: Why do we need to explicitly calculate the mask here? From the [docs](https://iris-esmf-regrid.readthedocs.io/en/stable/_api_generated/esmf_regrid.schemes.html#module-esmf_regrid.schemes), it looks like `use_src_mask` and `use_tgt_mask` can only be booleans. So why don't we just use `True/False` here?

It's also dangerous to derive this only from the first time step, as the mask might be time-dependent (e.g. for satellite data).

File: doc/recipe/preprocessor.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (bouweandela) [reply doc/recipe/preprocessor.rst:1005]: Yes, I was referring to "our" definition of "irregular". I now understand that that overlaps with UGRID cubes, but our definition would also apply to cubes that do not follow the UGRID standard, so would not have a mesh, right? And then iris-esmf-regrid only supports the ones that have a `.mesh` attribute that is not `None`.

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        given to ESMF for calculation. If resolution is set, the source cube
+        must have strictly increasing bounds (bounds may be transposed
+        plus or minus 360 degrees to make the bounds strictly increasing).
+        Only available for method 'conservative'.
+    tgt_resolution:
+        If present, represents the amount of latitude slices per target cell
+        given to ESMF for calculation. If resolution is set, the target cube
+        must have strictly increasing bounds (bounds may be transposed
+        plus or minus 360 degrees to make the bounds strictly increasing).
+        Only available for method 'conservative'.
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:83]: I don't really understand this, but it looks like you just copied this from the regridder. In case you have a better understanding, could you rephrase it? For me, a "cell" contains one latitude, so multiple latitudes per cell sounds awkward...

File: tests/unit/preprocessor/_regrid/test_regrid.py [hunk 8]
```diff
@@ -265,50 +267,126 @@ def make_test_cube(shape):
     return cube
 
 
-def test_rechunk_on_increased_grid():
-    """Test that an increase in grid size rechunks."""
-    with dask.config.set({'array.chunk-size': '128 M'}):
+def make_test_cube_unstructured(shape):
```
Comment (schlunma) [inline tests/unit/preprocessor/_regrid/test_regrid.py:270]: I think this defines an irregular cube (using [our definitions](https://esmvaltool--2457.org.readthedocs.build/projects/ESMValCore/en/2457/recipe/preprocessor.html#grid-types)).

File: tests/unit/preprocessor/_regrid/test_regrid.py [hunk 9]
```diff
@@ -265,50 +267,126 @@ def make_test_cube(shape):
...
+        iris.coords.AuxCoord(
+            np.broadcast_to(lon_points.reshape(1, -1), shape[-2:]),
+            standard_name='longitude',
+        ),
+        (-2, -1),
+    )
+    return cube
+
+
+def make_test_cube_irregular(shape):
```
Comment (schlunma) [inline tests/unit/preprocessor/_regrid/test_regrid.py:301]: I think this defines an unstructured cube (using [our definitions](https://esmvaltool--2457.org.readthedocs.build/projects/ESMValCore/en/2457/recipe/preprocessor.html#grid-types)).

**Review by schlunma — CHANGES_REQUESTED** [review]
Looks very good already Bouwe, it's great to see our custom ESMF code being replaced. Some additional comments:

1. In the PR description, please replace "unstructured" with "irregular" (you are replacing the schemes for irregular)
2. Would it make sense to deprecate our own `ESMFPy...` schemes? I don't think they are really useful anymore given that we have a much better solution for this now.
3. Would it maybe even make sense and go one step further by also replacing the schemes for regular regridding? This would make our code even easier. I think `esmf-regrid` also supports that kind of regridding with the same schemes. In principle, it would even work for unstructured grids, but in this case the data would need to be UGRID-compliant, and I am not sure we can enforce that (yet).

File: doc/recipe/preprocessor.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (bouweandela) [reply doc/recipe/preprocessor.rst:1005]: Oh wait, looks like I mixed up those definitions. In that case thuis text makes sense as it is.

[the remark this answers is not in the document]
Comment (schlunma) [reply doc/recipe/preprocessor.rst:1005]: As far as I understand, our definition of "irregular" does NOT overlap with UGRID. UGRID only refers to unstructured data (1D lat and lon with common dimension), whereas our definition of irregular corresponds to structured curvilinear data (2D lat and lon with common dimensions).

> And then iris-esmf-regrid only supports the ones that have a .mesh attribute that is not None.

This is only true for unstructured data, not for (our) irregular data.

File: esmvalcore/preprocessor/_regrid.py (hunk 1 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_regrid.py:544]: Yeah, that sounds like a good suggestion to me!

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 6 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:168]: It needs to be an array for data that contains a vertical dimension. This is allowed, as documented [here](https://iris-esmf-regrid.readthedocs.io/en/stable/_api_generated/esmf_regrid.schemes.html#esmf_regrid.schemes.ESMFAreaWeightedRegridder) and enforced by this code: https://github.com/SciTools-incubator/iris-esmf-regrid/blob/1772591d2ee1dbab06312e087ff694abe370cb72/esmf_regrid/schemes.py#L37-L76.

The mask seems mostly important to hide discontinuities in the grid. As long as the latitude and longitude coordinates are not dependent on time or vertical level, the mask computed here should be valid. It could mask out too much though, if fewer points are masked at a later timestep. However, computing the mask over all timesteps is computationally expensive as it requires loading the entire dataset from disk.

Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:168]: Are we currently supporting satellite datasets with time-varying masks?

File: tests/unit/preprocessor/_regrid/test_regrid.py (hunk 8 above)
Comment (bouweandela) [reply tests/unit/preprocessor/_regrid/test_regrid.py:270]: I mixed those two up

File: esmvalcore/preprocessor/_regrid.py (hunk 4 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid.py]: Done

File: doc/conf.py (hunk 3 above)
Comment (bouweandela) [reply doc/conf.py]: Done

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 7 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:83]: Yes, it's copy paste

File: tests/unit/preprocessor/_regrid/test_regrid.py (hunk 8 above)
Comment (bouweandela) [reply tests/unit/preprocessor/_regrid/test_regrid.py]: Fixed now

File: tests/unit/preprocessor/_regrid/test_regrid.py (hunk 9 above)
Comment (bouweandela) [reply tests/unit/preprocessor/_regrid/test_regrid.py]: Fixed now

**Comment by bouweandela** [comment]
Thanks for reviewing @schlunma! 

> In the PR description, please replace "unstructured" with "irregular" (you are replacing the schemes for irregular)

I mixed those two up, it should be better now.

> Would it make sense to deprecate our own ESMFPy... schemes? I don't think they are really useful anymore given that we have a much better solution for this now.

Yes, but maybe with the next release of iris-esmf-regrid where the `nearest` scheme is also supported. @stephenworsley mentioned that he expects to sort out https://github.com/SciTools-incubator/iris-esmf-regrid/pull/276 and https://github.com/SciTools-incubator/iris-esmf-regrid/pull/369 and make a new release soon.

> Would it maybe even make sense and go one step further by also replacing the schemes for regular regridding? This would make our code even easier. I think esmf-regrid also supports that kind of regridding with the same schemes. In principle, it would even work for unstructured grids, but in this case the data would need to be UGRID-compliant, and I am not sure we can enforce that (yet).

I like this idea, but there is an issue with ESMF that prevents parallelisation using Dask (see #2316). I have reported it to the ESMF developers and they said they saw opportunities to solve it, but I'm not sure when this will be solved. I would be keen to improve our parallisation scheme for those cases where ESMF is not needed and then it is relevant to only use it when is really needed.

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 6 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:168]: > Are we currently supporting satellite datasets with time-varying masks?

Yes, we have some in our data pool. It's also quite common that a mask is height-dependent, i.e., for lower levels the regions with high altitude are masked.

I assume the mask you can pass to the functions can only be 2D? This is really not optimal...it would be much simpler if iris-esmf supported masks with the same shape as the input data.

**Comment by schlunma** [comment]
Thanks for addressing my comments @bouweandela! Could you please add some documentation about the new schemes for meshes? It would also be great to use a common naming convention for this, e.g., at the moment, the documentation only talks about UGRID instead of iris meshes (in the [ICON doc](https://docs.esmvaltool.org/projects/ESMValCore/en/latest/quickstart/find_data.html#icon)). Maybe you could move that part to regridding docs and only reference that in the ICON-specific section?

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 6 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:168]: >  It's also quite common that a mask is height-dependent, i.e., for lower levels the regions with high altitude are masked.

The current code already takes care of that. I'll add something for the case where the mask is time varying.

> I assume the mask you can pass to the functions can only be 2D?

The mask is used to compute the regridding weights. Therefore the mask that is required describes the grid, so its dimensions need to match those of the grid.

**Comment by bouweandela** [comment]
> Would it make sense to deprecate our own ESMFPy... schemes? I don't think they are really useful anymore given that we have a much better solution for this now.

Done in 9aaf1b2

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 6 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py]: I added the option to specify along which axes the mask varies using the `collapse_src_mask_along_axes` and `collapse_tgt_mask_along_axis` arguments, see [here](https://esmvaltool--2457.org.readthedocs.build/projects/ESMValCore/en/2457/api/esmvalcore.regridding_schemes.html#esmvalcore.preprocessor.regrid_schemes.IrisESMFRegrid) for the documentation. Does it look better now?

**Comment by bouweandela** [comment]
@schlunma I've improved the documentation and testing and also added support for lazy nearest neighbour regridding now because this is at least partly supported with the new iris-esmf-regrid v0.11. Could you do another review, please?

File: esmvalcore/preprocessor/_regrid_esmpy.py [hunk 10]
```diff
@@ -161,6 +161,11 @@ class ESMPyAreaWeighted(_ESMPyScheme):
 
     Does not support lazy regridding.
 
+    .. deprecated:: 2.12.0
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid_esmpy.py:164]: Would it make sense to also print a warning if these schemes are used in a recipe? E.g., add an `ESMValCoreDeprecationWarning` to the `_ESMPyRegridder` code?

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+            coord = cube.coord(axis=axis, dim_coords=True)
+        except iris.exceptions.CoordinateNotFoundError:
+            try:
+                coord = cube.coord(axis=axis)
+            except iris.exceptions.CoordinateNotFoundError:
+                return tuple()
+        return cube.coord_dims(coord)
+
+    dims = {d for axis in axes for d in _get_dims_along_axis(cube, axis)}
+    return tuple(sorted(dims))
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:42]: I think this code could also added to the `preprocessor._shared` module. It's very general and could be useful for many preprocessors, not just regridding.

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+    collapse_src_mask_along_axes:
+        When deriving the mask from the source cube data, collapse the mask
+        along the dimensions identified by these axes. Only points that are
+        masked at all time (``'T'``), vertical levels (``'Z'``), or both time
+        and vertical levels (``'TZ'``) will be considered masked.
+    collapse_tgt_mask_along_axes:
+        When deriving the mask from the target cube data, collapse the mask
+        along the dimensions identified by these axes. Only points that are
+        masked at all time (``'T'``), vertical levels (``'Z'``), or both time
+        and vertical levels (``'TZ'``) will be considered masked.
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:95]: If an axis exists, but is not mentioned here, the first slice along that axis will be used right? Would be good to mention that here.

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+        which will tell :mod:`esmpy` which points to ignore. If an array is
+        provided, that will be used.
+        If set to :obj:`None`, it will be set to :obj:`True` for methods
+        ``'bilinear'`` and ``'conservative'`` and to :obj:`False` for method
+        ``'nearest'``. This default may be changed to :obj:`True` for all
+        schemes once `SciTools-incubator/iris-esmf-regrid#368
+        <https://github.com/SciTools-incubator/iris-esmf-regrid/issues/368>`_
+        has been resolved.
+    use_tgt_mask:
+        If True, derive a mask from (first time step) of the target cube,
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:78]:
```suggestion
        If True, derive a mask from of the target cube,
```
?

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py [hunk 14]
```diff
@@ -0,0 +1,233 @@
...
+        Only available for method 'conservative'.
+    tgt_resolution:
+        If present, represents the amount of latitude slices per target cell
+        given to ESMF for calculation. If resolution is set, the target cube
+        must have strictly increasing bounds (bounds may be transposed
+        plus or minus 360 degrees to make the bounds strictly increasing).
+        Only available for method 'conservative'.
+    tgt_location:
+        Either "face" or "node". Describes the location for data on the mesh
+        if the target is not a :class:`~iris.cube.Cube`.
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:110]: What happens if `tgt_location` is `None` for meshes?

**Review by schlunma — COMMENTED** [review]
Great! I just tested this extensively with all different grid types, and everything works as excepted 🚀 

Just have a couple of minor comments on the doc. Cheers!

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 12 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:95]: Yes. Do you think that is a safe option, also for ocean data? Or should we also allow e.g. coordinate names in addition to axes `'T'` and `'Z'`?

Comment (schlunma) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:76]: It makes certainly sense in some use cases, as long as it's documented properly. And yes, allowing coordinate names in addition to axes sounds great!

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 11 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py]: Done in 42123c0

File: esmvalcore/preprocessor/_regrid_esmpy.py (hunk 10 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_esmpy.py:182]: Done in 286f43d

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 14 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:91]: It looks like an exception will be raised: https://github.com/SciTools-incubator/iris-esmf-regrid/blob/9e2559af148b0e5ac057fcfeacac3a7ec3e18739/esmf_regrid/schemes.py#L498-L523

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 12 above)
Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py]: Done in cfc4330

**Comment by bouweandela** [comment]
Thanks for reviewing @schlunma. I think I have addressed all comments now, could you have another look please?

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py (hunk 14 above)
Comment (schlunma) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:102]: Ah, I was briefly confused why I didn't get an error in my tests when using the default scheme (which uses `tgt_location=None`), but since I used [unstructured->regular regridding](https://github.com/SciTools-incubator/iris-esmf-regrid/blob/9e2559af148b0e5ac057fcfeacac3a7ec3e18739/esmf_regrid/schemes.py#L1355), this keyword is ignored. I think it would be good to mention that this is required if target is a mesh, but ignored otherwise.

Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py]: Improved in a848468

File: esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+        plus or minus 360 degrees to make the bounds strictly increasing).
+        Only available for method 'conservative'.
+    tgt_resolution:
+        If present, represents the amount of latitude slices per target cell
+        given to ESMF for calculation. If resolution is set, the target cube
+        must have strictly increasing bounds (bounds may be transposed
+        plus or minus 360 degrees to make the bounds strictly increasing).
+        Only available for method 'conservative'.
+    tgt_location:
+        Only used if the target grid is an :class:`iris.mesh.MeshXY`. Describes
```
Comment (schlunma) [inline esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:102]:
```suggestion
        Only used if the target grid is an :class:`iris.mesh.MeshXY` or an :class:`iris.cube.Cube` with an associated mesh. Describes
```

Comment (bouweandela) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:102]: It is not used if the target is a cube (on a mesh), see the original docstring: https://iris-esmf-regrid.readthedocs.io/en/latest/_api_generated/esmf_regrid.schemes.html#esmf_regrid.schemes.ESMFBilinear

In that case, the `location` attribute of the cube is used: https://github.com/SciTools-incubator/iris-esmf-regrid/blob/9e2559af148b0e5ac057fcfeacac3a7ec3e18739/esmf_regrid/schemes.py#L504

Comment (schlunma) [reply esmvalcore/preprocessor/_regrid_iris_esmf_regrid.py:102]: Good point, haven't checked the code in that detail 😄

**Review by schlunma — APPROVED** [review]
Awesome, thanks @bouweandela! This will facilitate regridding enormously 🚀
